feat(cli): auto-detect language for single-language custom templates

rohang9000 opened the pull request:
Updated `cdk init` command to auto-detect language for single-language custom templates that are pulled from filesystem using `--from-path` so users do not need to specify `--language` in this case.

Fixes #

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

rohang9000 committed:
feat(cli): auto-detect language for single-language custom templates

iankhou reviewed (changes requested):
Please add unit tests for these changes. Consider separating the `basePath` and `actualBasePath` logic into a separate function, or adding some comments to explain it and its relationship with `language`—it's not straightforward at a glance.

iankhou committed:
Merge branch 'main' into fix/auto-detect-single-language-template

aws-cdk-automation committed:
chore(deps): upgrade dependencies (#839)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]:
https://github.com/aws/aws-cdk-cli/actions/runs/17625260603

------

*Automatically created by projen via the "upgrade" workflow*

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

aws-cdk-automation committed:
chore(deps): upgrade dependencies (#841)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]:
https://github.com/aws/aws-cdk-cli/actions/runs/17643754247

------

*Automatically created by projen via the "upgrade" workflow*

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

iliapolo committed:
chore: enable trusted publishing for releases to npm and pypi (#842)

This allows us to remove NPM and PyPI tokens from our secrets. Draft
until I also make the necessary changes in our NPM and PyPI accounts.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

iliapolo committed:
chore: npm trusted publishing for jsii packages is ignored (#843)

For jsii packages, the NPM publishing takes a different code path.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

aws-cdk-automation committed:
chore(deps): upgrade dependencies (#844)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]:
https://github.com/aws/aws-cdk-cli/actions/runs/17653263053

------

*Automatically created by projen via the "upgrade" workflow*

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

iliapolo committed:
chore: bump node (and npm) version in publish jobs to allow for trusted publishing (#845)

Trusted publishing is only supported starting from NPM 11.5.1 , which is
available in Node 24.x

> See
https://projen.io/docs/publishing/trusted-publishing/#meeting-the-npm-version-requirement

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

otaviomacedo committed:
fix(cli): stack filter positional arg is not being respected (#846)

At some point, the positional argument was lost, and there were no tests
for it.

Add it to `cli-config.ts`.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

aws-cdk-automation committed:
chore(deps): upgrade dependencies (#847)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]:
https://github.com/aws/aws-cdk-cli/actions/runs/17718646370

------

*Automatically created by projen via the "upgrade" workflow*

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

pahud committed:
chore: add close-stale-issues workflow (#849)

Adding `close-stale-issues` workflow just as [the
one](https://github.com/aws/aws-cdk/blob/main/.github/workflows/close-stale-issues.yml)
from aws/aws-cdk repo.

All configuration remains the same. Just changed the `repo name`.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

---------

Co-authored-by: Momo Kornher <mail@moritzkornher.de>

mrgrain committed:
chore: fix stale-issue-cleanup version (#851)

Explicitly define version. They've stopped publishing major version tag
for some reason.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

rix0rrr committed:
fix(cloud-assembly-schema): `unconfiguredBehavesLike` contains info for multiple versions (#782)

The way it's currently specified, the `unconfiguredBehavesLike` field
that's in the Cloud Assembly Schema requires the CLI to know the current
major version of the CDK Library to properly evaluate:

It looks like `unconfiguredBehavesLike: { v1: ..., v2: ... }` subfields,
and that requires the CLI to know whether to look in the `v1` or `v2`
subfield (and in the future `v3`, `v4`, etc...).

That is unnecessary: since the CDK Library knows its own current version
so it could just have communicated `unconfiguredBehavesLike: true` to
accurately reflect its current behavior, without the consumer needing to
know anything about the version.

The CLI code right now is only reading the `v2` field, but that will
become inaccurate if we ever release CDKv3.

In this PR, we're saying that the `v2` subfield is the "official"
location of this value going forward. `v1` should be ignored if present
(but is still included to not trip up any JSON Schema validation).

This leads to the least amount of interference with the existing
ecosystem (no forced upgrade by CDK users and no backwards compatibility
path in the CLI to potentially read 2 fields).

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

mrgrain committed:
chore(cli): remove deprecated AWS SDK v3 utilities in favor of Smithy equivalents (#852)

Removes left-over, long deprecated AWS SDK v3 utility packages. We are
already using the Smithy equivalents:

- Removes `@aws-sdk/middleware-endpoint` in favor of
`@smithy/middleware-endpoint`
- Removes `@aws-sdk/util-retry` in favor of `@smithy/util-retry`  
- Removes `@aws-sdk/util-waiter` in favor of `@smithy/util-waiter`

This change aligns with AWS SDK v3's migration path where utility
functions have been moved to the Smithy protocol layer. The
functionality remains the same as we were already using the newer,
non-deprecated packages.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

mrgrain committed:
chore: revert aws-actions/stale-issue-cleanup back to v6 (#853)

There are some weird issues with the latest version. Let's just go back
to the tried and tested on.


https://github.com/aws/aws-cdk-cli/actions/runs/17833245041/job/50705475999

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

mrgrain committed:
chore: fix codeowner team name (#856)

This was using an incorrect name. The correct name is
@aws/aws-cdk-core-team.

## Was

<img width="720" height="517" alt="image"
src="https://github.com/user-attachments/assets/8b329077-8d96-40e6-bcf2-2c14526b1b35"
/>

## Now

<img width="458" height="339" alt="image"
src="https://github.com/user-attachments/assets/ab854646-4086-4e53-8e68-64520dcf12d0"
/>

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

dgandhi62 committed:
refactor(cli): property bag for refreshStacks function (#855)

The refreshStacks function currently takes 4 individual parameters,
making it difficult to maintain as new parameters are added.

This change:
- Introduces a `RefreshStacksProps` interface to group function
parameters
- Refactors the `refreshStacks` function signature to accept a single
props object

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

Co-authored-by: Dhyan Gandhi <gandhya@amazon.com>
Co-authored-by: Kaizen Conroy <36202692+kaizencc@users.noreply.github.com>

aws-cdk-automation committed:
chore(deps): upgrade dependencies (#857)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]:
https://github.com/aws/aws-cdk-cli/actions/runs/17900936578

------

*Automatically created by projen via the "upgrade" workflow*

---------

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

iliapolo committed:
fix(cli-integ): sporadic failures due to external causes (#897)

Some of our tests are failing sporadically due to factors out of our
control. In the past week, here are the failing tests:

<img width="1249" height="216" alt="Screenshot 2025-09-28 at 10 33
31 PM"
src="https://github.com/user-attachments/assets/c19c7d40-ef1f-4e7a-ad96-887a9f2e1ec2"
/>

Causes I observed so far:

- Prolonged docker build durations, most likely caused by network or CPU
pressure since we are running a lot of parallel processes.
- Lambda functions fail to create due to IAM eventual consistency
issues.

To mitigate this, add a global timeout for tests to execute 3 times
before finally failing.

> See
https://jestjs.io/docs/jest-object#jestretrytimesnumretries-options

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

mrgrain committed:
feat(toolkit-lib): add forceDeployment option to BootstrapOptions (#898)

Fixes #891

This PR adds the `forceDeployment` property to the `BootstrapOptions`
interface to provide parity with the CLI `--force` flag. This allows
programmatic bootstrap operations to force re-bootstrap even when a
bootstrap stack already exists.

## Changes
- Added `forceDeployment?: boolean` property to `BootstrapOptions`
interface
- Added comprehensive test coverage for the new option
- Ensures the force option is properly passed through to SDK calls

## Use Cases
This enables:
- Management account bootstrapping with guaranteed up-to-date bootstrap
stacks
- CI/CD pipelines with clean bootstrap state guarantees  
- Multi-account deployment tools with flexible bootstrap behaviors

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

kaizencc committed:
chore(cli): add cdk flags to readme command list (#899)

the docs are already in the readme, just don't show up in the list

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

vivian12300 committed:
feat(cli): allow users to enable all feature flags that do not impact their application (#798)

Allows users to run `cdk flags --safe` to set all unconfigured feature
flags to their recommended states without changing their CloudFormation
template.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

---------

Co-authored-by: Rico Huijbers <rix0rrr@gmail.com>

mrgrain committed:
refactor(toolkit-lib): standardize confirmation requests to use ConfirmationRequest interface (#904)

This PR standardizes user confirmation flows across the CDK CLI by
migrating from ad-hoc `promptly.confirm` usage to the structured
`ConfirmationRequest` interface.

These are the changes as suggested here:
https://github.com/aws/aws-cdk-cli/pull/826#pullrequestreview-3283634674

## Changes

- **Convert CDK_TOOLKIT_I8910**: Changed from `DataRequest` to
`ConfirmationRequest` for refactor confirmation
- **Standardize destroy flow**: Replace `promptly.confirm` with
`IO.CDK_TOOLKIT_I7010` message
- **Standardize deploy flow**: Update `askUserConfirmation` to use
`ActionLessRequest<ConfirmationRequest>` pattern
- **Clean up test infrastructure**: Remove deprecated `markTesting`
function and `TESTING` variable
- **Update tests**: Replace `promptly` mocks with `requestSpy` for
consistency

## Benefits

- Consistent user interaction patterns across all CLI commands
- Better testability through standardized mocking
- Improved error handling and TTY detection through the IO system
- Cleaner separation of concerns between UI logic and business logic

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

mrgrain committed:
chore(cli-lib-alpha): stop releasing deprecated package (#905)

Stop releasing the deprecated `@aws-cdk/cli-lib-alpha` package. This
change removes the package from the code base and integ tests. If we
ever need to release it again, we can revert this PR.

I tried to just disable releasing, but then integ tests fail and I think
removing it all in one PR is just easier if we need to go back to it.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

mrgrain committed:
feat(integ-runner): use toolkit-lib as default engine (#906)

This PR makes the toolkit-lib engine the default for integ-runner,
replacing the previous cli-wrapper engine.

## Changes

- **Default engine switch**:  is now the default engine instead of 
- **Deprecation path**: Added unstable flag to temporarily revert to the
old engine
- **User guidance**: Added helpful error messages and deprecation
warnings to guide users through the transition
- **Test updates**: Updated test suite to reflect the new default engine

## Migration Path

Users experiencing issues with the new default engine can temporarily
revert using:
```
--unstable=deprecated-cli-engine
```

The old engine is scheduled for removal in January 2026.

## Error Handling

When errors occur with the new engine, users receive guidance on how to
report issues and temporarily revert if needed.

---
By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache-2.0 license

iankhou committed:
address my comments and increase test coverage for init.ts (97.81%)

iankhou commented on packages/aws-cdk/lib/commands/init/init.ts:
```diff
@@ -336,10 +354,33 @@ export class InitTemplate {
+
+    // Auto-detect single language templates
+    if (languages.length === 0) {
+      const entries = await fs.readdir(basePath, { withFileTypes: true });
+      const languageDirs = entries.filter(entry =>
+        entry.isDirectory() &&
+        ['typescript', 'javascript', 'python', 'java', 'csharp', 'fsharp', 'go'].includes(entry.name),
+      );
+
+      if (languageDirs.length === 1) {
```
Misread, this code is fine.

iankhou committed:
Merge branch 'main' into fix/auto-detect-single-language-template

iankhou commented:
Addressed my own comments and added a lot of unit testing to cover both new and old code.

Removed repeated code and consolidated the source of truth for supported CDK languages in a new `LanguageInfo` interface.

iankhou commented on packages/aws-cdk/lib/commands/init/init.ts:
```diff
@@ -336,10 +354,33 @@ export class InitTemplate {
+      const entries = await fs.readdir(basePath, { withFileTypes: true });
+      const languageDirs = entries.filter(entry =>
+        entry.isDirectory() &&
+        ['typescript', 'javascript', 'python', 'java', 'csharp', 'fsharp', 'go'].includes(entry.name),
```
Addressed this by implementing my suggestion. Moved language extensions to a new interface `LanguageInfo`.

iankhou commented on packages/aws-cdk/lib/commands/init/init.ts:
```diff
@@ -336,10 +354,33 @@ export class InitTemplate {
+
+        if (hasValidFiles) {
+          actualBasePath = path.join(basePath, langDir);
+          languages = [langDir];
```
Removed dead code.

iankhou commented:
[init.ts.html](https://github.com/user-attachments/files/22733286/init.ts.html)
